File: cli_anything/hotelbyte/commands/portal/entity.py
"""commands/portal/entity.py — tenant entity, customers, subscriptions, suppliers, settings."""
from __future__ import annotations

import json
from typing import Any

import click

from ...core.auth import authenticate_portal
from ...core.config import DEFAULT_ENV, load_profile
from ...core.http import HttpClient, HotelByteError
from ...utils.output import emit, error


def _client(ctx: click.Context) -> HttpClient:
    env = ctx.obj.get("env", DEFAULT_ENV)
    profile = load_profile("portal", env)
    if ctx.obj.get("username"):
        profile.username = ctx.obj["username"]
    if ctx.obj.get("password"):
        profile.password = ctx.obj["password"]
    authenticate_portal(profile)
    return HttpClient(profile)


# ── entity ─────────────────────────────────────────────────────────────

@click.group("entity")
def entity_group():
    """Tenant entity management (brands, groups, distribution config)."""


@entity_group.command("list")
@click.pass_context
def list_entities(ctx):
    """List entities accessible to the current user."""
    try:
        c = _client(ctx)
        resp = c.post("/api/user/tenant/listEntity", {})
        emit(resp, ctx.obj.get("json", False))
    except HotelByteError as e:
        error(str(e), ctx.obj.get("json", False))
        ctx.exit(1)


@entity_group.command("get")
@click.option("--entity-id", help="Entity ID (omit for current)")
@click.pass_context
def get_entity(ctx, entity_id):
    """Get a single entity."""
    body: dict[str, Any] = {}
    if entity_id:
        body["entityId"] = entity_id
    try:
        c = _client(ctx)
        resp = c.post("/api/user/tenant/getEntity", body)
        emit(resp, ctx.obj.get("json", False))
    except HotelByteError as e:
        error(str(e), ctx.obj.get("json", False))
        ctx.exit(1)


@entity_group.command("update")
@click.option("--entity-data", required=True, help="JSON entity update data, or @file.json")
@click.pass_context
def update_entity(ctx, entity_data):
    """Update an entity's configuration."""
    path = entity_data[1:] if entity_data.startswith("@") else entity_data
    try:
        body = json.loads(path)
    except json.JSONDecodeError:
        with open(path) as f:
            body = json.load(f)
    try:
        c = _client(ctx)
        resp = c.post("/api/user/tenant/updateEntity", body)
        emit(resp, ctx.obj.get("json", False))
    except HotelByteError as e:
        error(str(e), ctx.obj.get("json", False))
        ctx.exit(1)


@entity_group.command("distribution-config")
@click.option("--entity-id", help="Entity ID")
@click.pass_context
def distribution_config(ctx, entity_id):
    """Get effective distribution config for an entity."""
    body: dict[str, Any] = {}
    if entity_id:
        body["entityId"] = entity_id
    try:
        c = _client(ctx)
        resp = c.post("/api/user/tenant/getEffectiveDistributionConfig", body)
        emit(resp, ctx.obj.get("json", False))
    except HotelByteError as e:
        error(str(e), ctx.obj.get("json", False))
        ctx.exit(1)


@entity_group.command("policy-status")
@click.option("--entity-id", help="Entity ID")
@click.pass_context
def policy_status(ctx, entity_id):
    """Get entity policy status."""
    body: dict[str, Any] = {}
    if entity_id:
        body["entityId"] = entity_id
    try:
        c = _client(ctx)
        resp = c.post("/api/user/tenant/getEntityPolicyStatus", body)
        emit(resp, ctx.obj.get("json", False))
    except HotelByteError as e:
        error(str(e), ctx.obj.get("json", False))
        ctx.exit(1)


# ── customers ──────────────────────────────────────────────────────────

@click.group("customers")
def customers_group():
    """Customer lifecycle management."""


@customers_group.command("activate")
@click.option("--customer-id", required=True)
@click.pass_context
def activate_customer(ctx, customer_id):
    try:
        c = _client(ctx)
        resp = c.post("/api/user/tenant/activateCustomer", {"customerId": customer_id})
        emit(resp, ctx.obj.get("json", False))
    except HotelByteError as e:
        error(str(e), ctx.obj.get("json", False))
        ctx.exit(1)


@customers_group.command("inactivate")
@click.option("--customer-id", required=True)
@click.pass_context
def inactivate_customer(ctx, customer_id):
    try:
        c = _client(ctx)
        resp = c.post("/api/user/tenant/inactivateCustomer", {"customerId": customer_id})
        emit(resp, ctx.obj.get("json", False))
    except HotelByteError as e:
        error(str(e), ctx.obj.get("json", False))
        ctx.exit(1)


@customers_group.command("delete")
@click.option("--customer-id", required=True)
@click.pass_context
def delete_customer(ctx, customer_id):
    try:
        c = _client(ctx)
        resp = c.post("/api/user/tenant/deleteCustomer", {"customerId": customer_id})
        emit(resp, ctx.obj.get("json", False))
    except HotelByteError as e:
        error(str(e), ctx.obj.get("json", False))
        ctx.exit(1)


# ── subscriptions ───────────────────────────────────────────────────────

@click.group("subscriptions")
def subscriptions_group():
    """Subscription and billing management."""


@subscriptions_group.command("get")
@click.pass_context
def get_subscription(ctx):
    try:
        c = _client(ctx)
        resp = c.post("/api/user/tenant/getSubscription", {})
        emit(resp, ctx.obj.get("json", False))
    except HotelByteError as e:
        error(str(e), ctx.obj.get("json", False))
        ctx.exit(1)


@subscriptions_group.command("catalog")
@click.pass_context
def catalog(ctx):
    try:
        c = _client(ctx)
        resp = c.post("/api/user/tenant/getSubscriptionCatalog", {})
        emit(resp, ctx.obj.get("json", False))
    except HotelByteError as e:
        error(str(e), ctx.obj.get("json", False))
        ctx.exit(1)


@subscriptions_group.command("start")
@click.option("--plan-id", required=True)
@click.pass_context
def start_subscription(ctx, plan_id):
    try:
        c = _client(ctx)
        resp = c.post("/api/user/tenant/startSubscription", {"planId": plan_id})
        emit(resp, ctx.obj.get("json", False))
    except HotelByteError as e:
        error(str(e), ctx.obj.get("json", False))
        ctx.exit(1)


@subscriptions_group.command("change-plan")
@click.option("--plan-id", required=True)
@click.pass_context
def change_plan(ctx, plan_id):
    try:
        c = _client(ctx)
        resp = c.post("/api/user/tenant/changeSubscriptionPlan", {"planId": plan_id})
        emit(resp, ctx.obj.get("json", False))
    except HotelByteError as e:
        error(str(e), ctx.obj.get("json", False))
        ctx.exit(1)


@subscriptions_group.command("cancel")
@click.pass_context
def cancel_subscription(ctx):
    try:
        c = _client(ctx)
        resp = c.post("/api/user/tenant/cancelSubscription", {})
        emit(resp, ctx.obj.get("json", False))
    except HotelByteError as e:
        error(str(e), ctx.obj.get("json", False))
        ctx.exit(1)


@subscriptions_group.command("billing-portal")
@click.pass_context
def billing_portal(ctx):
    try:
        c = _client(ctx)
        resp = c.post("/api/user/tenant/getBillingPortal", {})
        emit(resp, ctx.obj.get("json", False))
    except HotelByteError as e:
        error(str(e), ctx.obj.get("json", False))
        ctx.exit(1)


@subscriptions_group.command("payment-methods")
@click.pass_context
def payment_methods(ctx):
    try:
        c = _client(ctx)
        resp = c.post("/api/user/tenant/listSubscriptionPaymentMethods", {})
        emit(resp, ctx.obj.get("json", False))
    except HotelByteError as e:
        error(str(e), ctx.obj.get("json", False))
        ctx.exit(1)


@subscriptions_group.command("invoices")
@click.pass_context
def invoices(ctx):
    try:
        c = _client(ctx)
        resp = c.post("/api/user/tenant/listSubscriptionInvoices", {})
        emit(resp, ctx.obj.get("json", False))
    except HotelByteError as e:
        error(str(e), ctx.obj.get("json", False))
        ctx.exit(1)


# ── suppliers ───────────────────────────────────────────────────────────

@click.group("suppliers")
def suppliers_group():
    """Supplier connection and credentials."""


@suppliers_group.command("accessible")
@click.pass_context
def accessible_credentials(ctx):
    try:
        c = _client(ctx)
        resp = c.post("/api/user/tenant/getAccessibleCredentials", {})
        emit(resp, ctx.obj.get("json", False))
    except HotelByteError as e:
        error(str(e), ctx.obj.get("json", False))
        ctx.exit(1)


@suppliers_group.command("connect")
@click.option("--supplier-id", required=True)
@click.option("--credentials", help="JSON credentials, or @file.json")
@click.pass_context
def connect_supplier(ctx, supplier_id, credentials):
    body: dict[str, Any] = {"supplierId": supplier_id}
    if credentials:
        path = credentials[1:] if credentials.startswith("@") else credentials
        try:
            body["credentials"] = json.loads(path)
        except json.JSONDecodeError:
            with open(path) as f:
                body["credentials"] = json.load(f)
    try:
        c = _client(ctx)
        resp = c.post("/api/user/tenant/connectSupplier", body)
        emit(resp, ctx.obj.get("json", False))
    except HotelByteError as e:
        error(str(e), ctx.obj.get("json", False))
        ctx.exit(1)


# ── retail onboarding ──────────────────────────────────────────────────

@click.group("retail")
def retail_group():
    """Retail onboarding status."""


@retail_group.command("status")
@click.pass_context
def retail_status(ctx):
    try:
        c = _client(ctx)
        resp = c.post("/api/user/tenant/getRetailOnboardingStatus", {})
        emit(resp, ctx.obj.get("json", False))
    except HotelByteError as e:
        error(str(e), ctx.obj.get("json", False))
        ctx.exit(1)

File: src/commands/portal/entity.ts
/**
 * commands/portal/entity.ts — entity, customers, subscriptions, suppliers, retail.
 */

import { Command } from "commander";
import { runPortal, type Ctx } from "./helpers.ts";
import { parseJsonInput } from "../../utils/output.ts";

// ── entity ─────────────────────────────────────────────────────────────

function createEntityGroup(ctx: Ctx): Command {
  const entity = new Command("entity").description("Tenant entity management (brands, groups, distribution config)");

  entity.command("list").description("List entities accessible to the current user")
    .action(async () => { await runPortal(ctx, "/api/user/tenant/listEntity", {}); });

  entity.command("get").description("Get a single entity")
    .option("--entity-id <id>", "Entity ID (omit for current)")
    .action(async (opts) => {
      const body: any = {};
      if (opts.entityId) body.entityId = opts.entityId;
      await runPortal(ctx, "/api/user/tenant/getEntity", body);
    });

  entity.command("update").description("Update an entity's configuration")
    .requiredOption("--entity-data <json>", "JSON entity update data, or @file.json")
    .action(async (opts) => {
      await runPortal(ctx, "/api/user/tenant/updateEntity", parseJsonInput(opts.entityData));
    });

  entity.command("distribution-config").description("Get effective distribution config for an entity")
    .option("--entity-id <id>", "Entity ID")
    .action(async (opts) => {
      const body: any = {};
      if (opts.entityId) body.entityId = opts.entityId;
      await runPortal(ctx, "/api/user/tenant/getEffectiveDistributionConfig", body);
    });

  entity.command("policy-status").description("Get entity policy status")
    .option("--entity-id <id>", "Entity ID")
    .action(async (opts) => {
      const body: any = {};
      if (opts.entityId) body.entityId = opts.entityId;
      await runPortal(ctx, "/api/user/tenant/getEntityPolicyStatus", body);
    });

  return entity;
}

// ── customers ──────────────────────────────────────────────────────────

function createCustomersGroup(ctx: Ctx): Command {
  const customers = new Command("customers").description("Customer lifecycle management");

  customers.command("activate").description("Activate a customer")
    .requiredOption("--customer-id <id>", "Customer ID")
    .action(async (opts) => { await runPortal(ctx, "/api/user/tenant/activateCustomer", { customerId: opts.customerId }); });

  customers.command("inactivate").description("Inactivate a customer")
    .requiredOption("--customer-id <id>", "Customer ID")
    .action(async (opts) => { await runPortal(ctx, "/api/user/tenant/inactivateCustomer", { customerId: opts.customerId }); });

  customers.command("delete").description("Delete a customer")
    .requiredOption("--customer-id <id>", "Customer ID")
    .action(async (opts) => { await runPortal(ctx, "/api/user/tenant/deleteCustomer", { customerId: opts.customerId }); });

  return customers;
}

// ── subscriptions ───────────────────────────────────────────────────────

function createSubscriptionsGroup(ctx: Ctx): Command {
  const subs = new Command("subscriptions").description("Subscription and billing management");

  subs.command("get").description("Get current subscription")
    .action(async () => { await runPortal(ctx, "/api/user/tenant/getSubscription", {}); });

  subs.command("catalog").description("List subscription catalog")
    .action(async () => { await runPortal(ctx, "/api/user/tenant/getSubscriptionCatalog", {}); });

  subs.command("start").description("Start a subscription")
    .requiredOption("--plan-id <id>", "Plan ID")
    .action(async (opts) => { await runPortal(ctx, "/api/user/tenant/startSubscription", { planId: opts.planId }); });

  subs.command("change-plan").description("Change subscription plan")
    .requiredOption("--plan-id <id>", "Plan ID")
    .action(async (opts) => { await runPortal(ctx, "/api/user/tenant/changeSubscriptionPlan", { planId: opts.planId }); });

  subs.command("cancel").description("Cancel subscription")
    .action(async () => { await runPortal(ctx, "/api/user/tenant/cancelSubscription", {}); });

  subs.command("billing-portal").description("Get billing portal URL")
    .action(async () => { await runPortal(ctx, "/api/user/tenant/getBillingPortal", {}); });

  subs.command("payment-methods").description("List subscription payment methods")
    .action(async () => { await runPortal(ctx, "/api/user/tenant/listSubscriptionPaymentMethods", {}); });

  subs.command("invoices").description("List subscription invoices")
    .action(async () => { await runPortal(ctx, "/api/user/tenant/listSubscriptionInvoices", {}); });

  return subs;
}

// ── suppliers ───────────────────────────────────────────────────────────

function createSuppliersGroup(ctx: Ctx): Command {
  const suppliers = new Command("suppliers").description("Supplier connection and credentials");

  suppliers.command("accessible").description("List accessible supplier credentials")
    .action(async () => { await runPortal(ctx, "/api/user/tenant/getAccessibleCredentials", {}); });

  suppliers.command("connect").description("Connect a supplier")
    .requiredOption("--supplier-id <id>", "Supplier ID")
    .option("--credentials <json>", "JSON credentials, or @file.json")
    .action(async (opts) => {
      const body: any = { supplierId: opts.supplierId };
      if (opts.credentials) body.credentials = parseJsonInput(opts.credentials);
      await runPortal(ctx, "/api/user/tenant/connectSupplier", body);
    });

  return suppliers;
}

// ── retail ──────────────────────────────────────────────────────────────

function createRetailGroup(ctx: Ctx): Command {
  const retail = new Command("retail").description("Retail onboarding status");
  retail.command("status").description("Get retail onboarding status")
    .action(async () => { await runPortal(ctx, "/api/user/tenant/getRetailOnboardingStatus", {}); });
  return retail;
}

// ── export all ──────────────────────────────────────────────────────────

export {
  createEntityGroup,
  createCustomersGroup,
  createSubscriptionsGroup,
  createSuppliersGroup,
  createRetailGroup,
};